tasks/components.py

```diff
@@ -160,116 +161,217 @@ def has_type_component(content) -> bool:
+    root_path = pathlib.Path(comp.path)
+
+    # Definition file `def/component.go` must define a component interface
+    if not any(
+        line.startswith('type Component interface') or line.startswith('type Component = ') for line in def_content
+    ):
+        return f"** {comp.def_file} does not define a Component interface"
+
+    # Skip components that need to migrate
+    if str(comp.def_file) in components_to_migrate:
+        return
+
+    # Special case for api
+    if comp.def_file == 'comp/api/api/def/component.go':
+        return
+
+    # Definition file `def/component.go` must use `package <compname>`
+    pkgname = parse_package_name(comp.def_file)
+    if pkgname != comp.name:
+        if comp.path not in ignore_package_name:
+            return f"** {comp.def_file} has wrong package name '{pkgname}', must be '{comp.name}'"
+
+    # Definition file `def/component.go` must not contain a mock definition
+    for mock_definition in mock_definitions:
+        if any(line.startswith(mock_definition) for line in def_content):
+            return f"** {comp.def_file} defines '{mock_definition}' which should be in separate implementation. See docs/components/defining-components.md"
+
+    # Allowlist of components that do not use an implementation folder
+    if comp.path in components_missing_implementation_folder:
+        return
+
+    # Implementation folder or folders must exist
+    impl_folders = locate_implementation_folders(comp)
+    if len(impl_folders) == 0:
+        return f"** {comp.name} is missing the implementation folder in {comp.path}. See docs/components/defining-components.md"
+
+    if comp.version == 2:
+        # Implementation source files should use correct package name, and shouldn't import fx (except tests)
+        for src_file in locate_nontest_source_files(impl_folders):
+            pkgname = parse_package_name(src_file)
+            expectname = comp.name + 'impl'
+            if pkgname != expectname:
+                if comp.path not in ignore_package_name:
+                    return f"** {src_file} has wrong package name '{pkgname}', must be '{expectname}'"
+            src_content = read_file_content(src_file)
+            if 'go.uber.org/fx' in src_content:
+                return f"** {src_file} should not import 'go.uber.org/fx' because it a component implementation"
+            if 'fxutil' in src_content:
+                return f"** {src_file} should not import 'fxutil' because it a component implementation"
+        # FX files should use correct package name
+        for src_file in locate_fx_source_files(root_path):
+            pkgname = parse_package_name(src_file)
+            expectname = comp.name + 'fx'
+            if pkgname != 'fx' and pkgname != expectname:
+                if comp.path not in ignore_package_name:
+                    return f"** {src_file} has wrong package name '{pkgname}', must be 'fx' or '{expectname}'"
```

Comment:
Should we check we are using the `fxutil.ProvideComponentConstructor` function in the `fx` folder?

Reply:
Good call, added a test for that as well.